### ISSUE TITLE
Initial render is painfully slow

### ISSUE THREAD

**Issue by SoonDead** [erikras/redux-form#3606]
### Are you submitting a **bug report** or a **feature request**?

Bug report

### What is the current behavior?

I'm in the process of rewriting an application that used 5.3.3 which used an older version of React. The new app is not a direct port but features very similar screens to the original.
I'm making a form where the form is an array of elements.
Rows are FieldArrays and can be added, removed, ordered. Each row contains 11 Fields. 
I initialize the form with server side data so the initial render contains 36 rows.
All in all there is around ~400 fields.
The initial render takes about 6 seconds while the browser hangs.

### What is the expected behavior?

Connecting 400 elements should be relatively fast. The whole premise of the v6 rework was performance of big forms.
I have a version of more or less the same form in v5.3.3 and it renders under 1 second on the same machine in the same browser.

Is this expected?

### What's your environment?

Chrome, React 16.0, Redux 3.6.0, ReactRedux: 5.0.5, ReduxForm: 7.1.2

### Other information

![image](https://user-images.githubusercontent.com/988167/32702331-c899a412-c7e5-11e7-8e73-9185a7f68f05.png)

**Comment by gustavohenke** [comment]
Well, the fact that v5 had an initial render faster than v6+ doesn't mean that subsequent re-renders (eg while typing) were faster as well, right? 😛  I haven't used v5 but AFAIK, this is where v6 wins.

Specific use cases like yours with so many fields are quite rare, but they do exist; we have some similar issues.

The best we can do at the moment is debug an example of your v6/v7 code in order to find any potential bottlenecks.
Also, if you have the know-how to dig into Redux Form and find ways to improve the performance, that would be great! We will welcome any contributions.

**Comment by SoonDead** [comment]
>> Well, the fact that v5 had an initial render faster than v6+ doesn't mean that subsequent re-renders (eg while typing) were faster as well, right? 😛 I haven't used v5 but AFAIK, this is where v6 wins.

Just trying out the old form now, it has no performance problems whatsoever. Typing, changing values has no lag. 400 fields is not that big of a form.

The slowest part in the flamechart seems to be connecting `<Fields />` (about half the time). I might be misusing it. What I need is to show/hide/reset/change some form elements depending on others.

I have the following form elements in each row:

```
isComposite: checkbox
unitType: select
unit: select
unitType2: select
unit2: select
```
Rules:
 * unitType2 and unit2 should be only visible if isComposite is true.
 * unit and unit2 has different values selectable according to unitType and unitType2 (different options)
 * therefore if unitType or unitType2 changes, the unit values should be also changed.

I have made a `<Fields />` element for the 5 values and show/hide specific values that single component.

What would be the intended way to do this?

**Comment by SoonDead** [comment]
I have refactored this to use `formValues`.

The slow part seems to be `Field.componentWillMount` which now takes 20-30ms for each field:

![image](https://user-images.githubusercontent.com/988167/32779684-feb36420-c93e-11e7-84d7-626cb6ff148a.png)

Drilling down into the main thread it seems that what takes up time is `detectMutations` and `trackForMutations` (and `trackProperties` inside) which both are 5-6ms each time.

![image](https://user-images.githubusercontent.com/988167/32779703-0decd8f4-c93f-11e7-8895-513e214af34a.png)

**Comment by SoonDead** [comment]
Oh for crying out loud, it was so slow because of the redux-immutable-state-invariant middleware in the store. Now it's back to perfectly acceptable speeds.

**Comment by gustavohenke** [comment]
So, with v5 your form was performing okay even though you used the redux-immutable-state-invariant middleware?

**Comment by SoonDead** [comment]
No, the bootstrapping of the previous version of the project was fundamentally different and redux-immutable-state-invariant was not used.
